### DISCUSSION TITLE
Animate ListBox item add from ViewModel

### DISCUSSION THREAD

**Discussion by workgroupengineering** [AvaloniaUI/Avalonia#11256] [Q&A]
How can i animate with reverse fade effect when item add from ViewModel?

Version: **11.0.0-Preview7**

**Comment by maxkatz6** [comment]
First idea I have in my head would be to handle container related events in ListBox. And when container is attached to the tree, run opacity animation (or change opacity property with transition set before).
But containers can be reused, which should be kept in mind.

**Reply by workgroupengineering** [reply]
Thank you for your suggestion. Can you get me to some documentation on how container recycling works?

**Reply by maxkatz6** [reply]
We probably don't have one, but generally it's just these intuitive methods:
```
ItemsControl.IsItemItsOwnContainerOverride
ItemsControl.PrepareContainerForItemOverride
ItemsControl.ContainerIndexChangedOverride
ItemsControl.ClearContainerForItemOverride
```

Where you probably need to override PrepareContainerForItemOverride and ClearContainerForItemOverride calling base implementation, but adding your own animation to the base container.
Alternatively, you can get ItemsControl.Generator and subscribe on the same events.

Some details also here in the PR description https://github.com/AvaloniaUI/Avalonia/pull/9677

**Reply by maxkatz6** [reply]
I also don't think WPF has detailed documentation for similar events.
